teach systemd-dissect file copying, and make it officially supported, move to /usr/bin + man page

poettering opened the pull request.

keszybz commented on TODO:
```diff
@@ -34,9 +34,6 @@ Features:
   fstype-agnostic version of btrfs' rootflags=subvol=foobar.
 
-* add --copy-from and --copy-to command to systemd-dissect which copies stuff
-  in and out of a disk image
```
Hmm, `--copy-{from,to}` are more self-explantory than `--extract/--add`. Esp. "add" is very generic... Do we have some "prior art" for the naming of this type of operations?

poettering replied:
I used "tar" as inspiration, i.e. "tar --extract" is the long form of "x" and "tar --add-file=FILE" is for adding in stuff...

poettering replied:
The "--copy-from=" and "--copy-to=" stuff I didn#t like to much since it wasn't clear what it copies from and to.

poettering commented on src/dissect/dissect.c:
```diff
@@ -370,6 +432,170 @@ static int run(int argc, char *argv[]) {
                 break;
 
+        case ACTION_EXTRACT:
+        case ACTION_ADD: {
```
I'll add a commit on top that splits the switch statement into three separate functions.

poettering commented on src/dissect/dissect.c:
```diff
@@ -327,18 +327,36 @@ static int run(int argc, char *argv[]) {
+                        arg_root_hash_sig_path || arg_root_hash_sig ? NULL : &arg_root_hash_sig_path);
         if (r < 0)
-                return log_error_errno(r, "Failed to set up loopback device: %m");
+                return log_error_errno(r, "Failed to read verity artefacts for %s: %m", arg_image);
```
oh, i just moved that around, didn't change the spelling. will fix though

poettering commented on src/dissect/dissect.c:
```diff
@@ -370,6 +432,170 @@ static int run(int argc, char *argv[]) {
+
+                        (void) copy_xattr(source_fd, target_fd);
+                        (void) copy_access(source_fd, target_fd);
+                        (void) copy_times(source_fd, target_fd, 0);
```
Hmm, so I though about this, and I am very sure we should not copy owners, because that clearly makes only sense within an image's context, hence I didn't do that. But whether a file is RO or not and auxiliary info in xattrs I figured is something that shouldn't be lost on the way... I mean, copying a file like /etc/shadow from an image to the host should probably not open it up to everyone, but if that's desired require a separate step...

Noone knows what xattrs contain, but I'd assume that the user xattrs (which is what this is about) are generally user meta information, and there's little reason to belive it makes no sense outside of the image?

bluca replied:
copying xattrs also is quite handy when a file is already selinux-labeled, and you want to avoid having to re-label it after the fact

poettering commented:
Hmm, any chance you can get me an "strace" of that systemd-dissect run on /dev/sda? I'd like to see where the ELOOP comes from...

poettering commented:
Force pushed a new (rebased) version. Addressed all @keszybz points, except for the points discussed above.

keszybz commented:
<details>

```console
$ sudo SYSTEMD_LOG_LEVEL=debug strace -ff build/systemd-dissect /dev/sda
...
getpid()                                = 115935
mkdir("/tmp/dissect-pxBEJC", 0700)      = 0
rt_sigprocmask(SIG_SETMASK, ~[RTMIN RT_1], [CHLD], 8) = 0
clone(child_stack=NULL, flags=CLONE_NEWNS|SIGCHLDstrace: Process 115936 attached
) = 115936
[pid 115935] writev(2, [{iov_base="Successfully forked off '(sd-dis"..., iov_len=53}, {iov_base="\n", iov_len=1}], 2Successfully forked off '(sd-dissect)' as PID 115936.
 <unfinished ...>
[pid 115936] prctl(PR_SET_NAME, "(sd-dissect)" <unfinished ...>
[pid 115935] <... writev resumed>)      = 54
[pid 115936] <... prctl resumed>)       = 0
[pid 115935] rt_sigprocmask(SIG_SETMASK, [CHLD],  <unfinished ...>
[pid 115936] geteuid( <unfinished ...>
[pid 115935] <... rt_sigprocmask resumed>NULL, 8) = 0
[pid 115936] <... geteuid resumed>)     = 0
[pid 115935] close(5 <unfinished ...>
[pid 115936] mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0 <unfinished ...>
[pid 115935] <... close resumed>)       = 0
[pid 115936] <... mmap resumed>)        = 0x7f1c516f6000
[pid 115935] fcntl(3, F_GETFL <unfinished ...>
[pid 115936] prctl(PR_SET_MM, PR_SET_MM_ARG_START, 0x7f1c516f6000, 0, 0 <unfinished ...>
[pid 115935] <... fcntl resumed>)       = 0 (flags O_RDONLY)
[pid 115936] <... prctl resumed>)       = 0
[pid 115935] fstat(3,  <unfinished ...>
[pid 115936] prctl(PR_SET_MM, PR_SET_MM_ARG_END, 0x7f1c516f600d, 0, 0 <unfinished ...>
[pid 115935] <... fstat resumed>{st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
[pid 115936] <... prctl resumed>)       = 0
[pid 115935] read(3,  <unfinished ...>
[pid 115936] prctl(PR_SET_PDEATHSIG, SIGTERM) = 0
[pid 115936] rt_sigaction(SIGHUP, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGINT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGQUIT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGILL, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGTRAP, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGABRT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGBUS, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGFPE, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGUSR1, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGSEGV, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGUSR2, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGPIPE, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGALRM, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGTERM, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGSTKFLT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGCHLD, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGCONT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGTSTP, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGTTIN, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGTTOU, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGURG, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGXCPU, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGXFSZ, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGVTALRM, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGPROF, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGWINCH, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGIO, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGPWR, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGSYS, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_2, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_3, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_4, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_5, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_6, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_7, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_8, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_9, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_10, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_11, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_12, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_13, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_14, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_15, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_16, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_17, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_18, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_19, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_20, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_21, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_22, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_23, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_24, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_25, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_26, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_27, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_28, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_29, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_30, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_31, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigaction(SIGRT_32, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1c51197a70}, NULL, 8) = 0
[pid 115936] rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
[pid 115936] getppid()                  = 115935
[pid 115936] mount(NULL, "/", NULL, MS_REC|MS_SLAVE, NULL) = 0
[pid 115936] getpid()                   = 115936
[pid 115936] writev(2, [{iov_base="Failed to mount dissected image:"..., iov_len=66}, {iov_base="\n", iov_len=1}], 2Failed to mount dissected image: Too many levels of symbolic links
) = 67
[pid 115936] exit_group(1)              = ?
[pid 115935] <... read resumed>"", 4096) = 0
[pid 115936] +++ exited with 1 +++
writev(2, [{iov_base="Failed to read /etc/hostname: No"..., iov_len=55}, {iov_base="\n", iov_len=1}], 2Failed to read /etc/hostname: No such file or directory
) = 56
close(3)                                = 0
close(7)                                = 0
fcntl(6, F_GETFL)                       = 0 (flags O_RDONLY)
fstat(6, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
read(6, "", 4096)                       = 0
writev(2, [{iov_base="/etc/machine-id file is empty.", iov_len=30}, {iov_base="\n", iov_len=1}], 2/etc/machine-id file is empty.
) = 31
close(6)                                = 0
close(9)                                = 0
fcntl(8, F_GETFL)                       = 0 (flags O_RDONLY)
fstat(8, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
fstat(8, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
read(8, "", 4096)                       = 0
close(8)                                = 0
close(11)                               = 0
fcntl(10, F_GETFL)                      = 0 (flags O_RDONLY)
fstat(10, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
fstat(10, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
read(10, "", 4096)                      = 0
close(10)                               = 0
waitid(P_PID, 115936, {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=115936, si_uid=0, si_status=1, si_utime=0, si_stime=0}, WEXITED, NULL) = 0
writev(2, [{iov_base="(sd-dissect) failed with exit st"..., iov_len=39}, {iov_base="\n", iov_len=1}], 2(sd-dissect) failed with exit status 1.
) = 40
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=115936, si_uid=0, si_status=1, si_utime=0, si_stime=0} ---
rmdir("/tmp/dissect-pxBEJC")            = 0
writev(2, [{iov_base="Failed to acquire image metadata"..., iov_len=48}, {iov_base="\n", iov_len=1}], 2Failed to acquire image metadata: Protocol error
) = 49
fsync(4)                                = 0
close(4)                                = 0
exit_group(1)                           = ?
+++ exited with 1 +++
```

I don't see why this is failing, but it's late here and I have been starting for the screen for way too long, so I might be missing something obvious...
</details>

poettering commented:
Ah, I figured it out. You have an encrypted root fs, right?

So this happens if we try to mount a file system without decrypting it first. It's the ELOOP returned by mount_partition() if the fs type is `crypto_LUKS`. The comment adjacent to that is a bit confusing, it assumes this is called when things were already decrypted, which would then mean things are "doubly" encrypted...

will improve

poettering commented:
Force pushed a new version with three new commits on top which improve error message generation. @keszybz' encrypted hard disk case should not result in a lovely error message. ptal!

bluca commented on src/shared/dissect-image.c:
```diff
@@ -1051,6 +1052,12 @@ static int mount_partition(
                         return -ENOMEM;
 
+        if (FLAGS_SET(flags, DISSECT_IMAGE_MKDIR)) {
+                r = mkdir_p(p, 0755);
```
in src/core/mount.c the target directory is created with the selinux version (mkdir_p_label) - is there any specific logic to when one is used vs the other?

poettering replied:
basically: unless it's really really obvious that we want selinux labelling we don't do it. and wait until some selinux people complain and then add it. Thing is, we don't maintain the policy, and we don't really know if the want to label something in some specific way or are fine with default labels.

Moreover, this stuff is relative to some image, and that image if it has selinux has likely a different policy applied, hence if we really tried to label things correctly here, we'd have to load the policy from the image, and selinux has no recursive policy loading/namespacing, hence we can't do it even if we would want to...

bluca replied:
Right, I've been pestering our SELinux folks for quite some time about namespaced policies, for this and other reasons. My plan is to keep annoying them until they implement it just to shut me up :-D
More seriously, this one is probably going to be one of the use cases where we need the labelling to happen. I'll ping them and follow up personally if needed, once we figure it out.

poettering replied:
@bluca note that this is only a fallback anyway, if the mount points to overmount are missing so far. Any properly build image should not need this, and the mkdir_p() should be a NOP.

I mostly added this, so that I can use systemd-repart to generate an image with initialized file systems but zero files in them, and still mount it with "systemd-dissect --mount". i.e. I wanted to be able to take an image consistent of a few partitions fresh out of "mkfs" and friends, mount it with `systemd-dissect --mount` and then rsync a file tree into it, to populate it. But given that mkfs leaves entirely empty fs by default, the dirs to mount over are not going to exist either in this case, so I figured, let's just create them — if they are missing.

A proper, ppoulated, selinux enabled image doesn't need any of this: the mkdir_p() should be a NOP there, because the mount point dirs would already exist.

Hence, I think it's totally fine not to do the selinux dance here: it's a fallback that is only useful in very specific cases, where you start from scratch. in the normal case it should not be needed, and if you have a problem with the labels it puts on stuff then the right fix is probably to create and label the mountpoint dirs beforehand properly, so that `systemd-dissect --mount` doesn't have to anymore.

bluca commented on src/dissect/dissect.c:
```diff
@@ -370,6 +432,170 @@ static int run(int argc, char *argv[]) {
+
+                r = dissected_image_decrypt_interactively(
+                                m, NULL,
+                                arg_root_hash, arg_root_hash_size,
```
If the verity metadata is passed through, ACTION_ADD would fail as the images are mounted read-only, right? So three options I can see:

1) Print a warning if any of those vars are set and ignore them
2) Print an error if ACTION_ADD and those vars are set and bail out
3) Stretch goal: use libcryptsetup to recreate the verity data after adding the file (quite a lot of work)

bluca replied:
Yep:

```
$ sudo buildx86/systemd-dissect img.raw
      Name: img.raw
      Size: 4.0K

Machine ID: 3d8a4aab19c64398aa48782f0a5e4ce8
OS Release: PRETTY_NAME=Debian GNU/Linux 10 (buster)
            NAME=Debian GNU/Linux
            VERSION_ID=10
            VERSION=10 (buster)
            VERSION_CODENAME=buster
            ID=debian
            HOME_URL=https://www.debian.org/
            SUPPORT_URL=https://www.debian.org/support
            BUG_REPORT_URL=https://bugs.debian.org/

RW DESIGNATOR PARTITION UUID FSTYPE   ARCHITECTURE VERITY   NODE    PARTNO
ro root       -              squashfs -            external img.raw      -
$ sudo buildx86/systemd-dissect --add img.raw /tmp/file /file
Failed to open target file '/file': Read-only file system
```

poettering replied:
well, you cannot add a file to a squashfs anyway. I think the error you get is actually pretty explanatory, no? it say the fs is read-only. Which it is, and it doesn't matter too much if that's because of verity or because squashfs is used or both.

I think the current behaviour makes a ton of sense: try it, and if it fails, log a good error about it, and i think the one you see is good.

bluca replied:
Aye, fair enough, the error is clear. I might have a use case where it would be handy to regenerate stuff - I have to repack squashfs/veritysetup it/sign it way too often, so I was going at it from that angle - having a nice tool that understands it would be cool. But I'll look into it myself, and see if it makes sense.

bluca commented on src/shared/dissect-image.c:
```diff
@@ -1840,7 +1906,16 @@ int dissected_image_acquire_metadata(DissectedImage *m) {
+                return r;
+
+        k = read(error_pipe[0], &v, sizeof(r));
+        if (k < 0)
```
CI is unhappy:

```
15:42:32 ../src/shared/dissect-image.c: In function ‘dissected_image_acquire_metadata’:
15:42:32 ../src/shared/dissect-image.c:1912:9: error: comparison of unsigned expression < 0 is always false [-Werror=type-limits]
15:42:32          if (k < 0)
15:42:32          ^
```

bluca commented:
The new output is nice, but test/units/testsuite-50.sh needs an update since it parses it - also, would be good to add test coverage for the new parameters

bluca commented on src/dissect/dissect.c:
```diff
@@ -207,139 +273,438 @@ static int parse_argv(int argc, char *argv[]) {
+        assert(m);
+        assert(d);
+
+        printf("      Name: %s\n", basename(arg_image));
```
indentation looks a bit strange with name/size more indented than the rest?

```
$ sudo buildx86/systemd-dissect img.raw
      Name: img.raw
      Size: 4.0K

Machine ID: 3d8a4aab19c64398aa48782f0a5e4ce8
OS Release: PRETTY_NAME=Debian GNU/Linux 10 (buster)
            NAME=Debian GNU/Linux
            VERSION_ID=10
            VERSION=10 (buster)
            VERSION_CODENAME=buster
            ID=debian
            HOME_URL=https://www.debian.org/
            SUPPORT_URL=https://www.debian.org/support
            BUG_REPORT_URL=https://bugs.debian.org/

RW DESIGNATOR PARTITION UUID FSTYPE   ARCHITECTURE VERITY   NODE    PARTNO
ro root       -              squashfs -            external img.raw      -
```

poettering replied:
what do you mean? everything above the partition table  is aligned along the `:`, like we usually do it.

bluca replied:
Ah I see - I guess my brain was expecting to see Name/Size left-aligned. If that's how all the tables are printed then it's fine.

bluca commented:
No strong opinion, but would the parameters maybe look better and match the look&feel of other tools as verbs? Eg: systemd-dissect --mount -> systemd-dissect mount

bluca commented:
Apart from those minor comments, looks really nice - thanks!

poettering commented:
> No strong opinion, but would the parameters maybe look better and match the look&feel of other tools as verbs? Eg: systemd-dissect --mount -> systemd-dissect mount

So, I thought about this. We could still change this, given that the tool so far is not installed to bindir. But how would we call the current default operation, where we dissect things? would that then be `systemd-dissect dissect foobar.raw`? that would be hard to read, and "stutter", if you understand what i mean. We could of course then also rename the tool to `systemd-image` or so, but I am not sure I like that too much.

I think as long as the primary job is showing dissected info, and the mount and copy stuff is just secondary we should live it as it is. And I think that's the case. does that make sense?

bluca commented:
> > No strong opinion, but would the parameters maybe look better and match the look&feel of other tools as verbs? Eg: systemd-dissect --mount -> systemd-dissect mount
> 
> So, I thought about this. We could still change this, given that the tool so far is not installed to bindir. But how would we call the current default operation, where we dissect things? would that then be `systemd-dissect dissect foobar.raw`? that would be hard to read, and "stutter", if you understand what i mean. We could of course then also rename the tool to `systemd-image` or so, but I am not sure I like that too much.
> 
> I think as long as the primary job is showing dissected info, and the mount and copy stuff is just secondary we should leave it as it is I think. And I think that's the case. does that make sense?

I was thinking the default would be no-verb - just like systemctl by default prints all the statuses, this by default prints all the information. It could have an alias as "info" for a verb.
But again no strong opinion either way - just a suggestion.

poettering commented:
> I was thinking the default would be no-verb - just like systemctl by default prints all the statuses, this by default prints all the information. It could have an alias as "info" for a verb.

Well, but how would you specify the image name then? 

right now `systemd-dissect foo.raw` shows you image info. If we switch to a verb based syntax, then there are two options:

1. `foo.raw` would have to be specified as `--image=foo.raw` or so, which would suck
2. or for the main operation, i.e. showing contents we'd have to have a verb to avoid ambiguity between what the verb is and what the image name.

or to say this differently: as I understand you want that the command lines look like this:

```
systemd-dissect $VERB $IMAGE
```

but then you want to make `$VERB` optional, so that this becomes equivalent:

```
systemd-dissect $IMAGE
systemd-dissect info $IMAGE
```

but then things become nasty if you want to inspect an image called "info", because `systemd-dissect info` could then mean two things: you want to see the data about info, or you mean the info verb but forgot the image name....

bluca commented:
All fair points - let's keep it as is, as said no strong opinion :)

bluca commented:
> The new output is nice, but test/units/testsuite-50.sh needs an update since it parses it - also, would be good to add test coverage for the new parameters

Just the tests to fix, then it LGTM!

poettering commented:
Pushed a new version now. I made some changes since the last version:

1. After all I followed @keszybz' suggestion to call the switches --copy-from/-copy-to instead of --extract/--add, because I realized we have prior art here: machinectl has verbs like that. And I figure consistency with ourselves matters. Hence I figured we should use the same nomenclature
2. I updated the dissection test, as @bluca pointed out. The tests check the output of the dissect tool, which is a lot more dynamic now that we use the table formatter. So in order to make this robust I did:
3. I added JSON output as an alternative, similar to how we have it now in various other tools. This should be helpful in various usecases, but also allows us to make the tests more reliable, since we don't compar human readable pretty strings anymore. As noted above.

No other changes!

PTAL!

bluca commented:
Checked json output on my machine, looks very nice! LGTM

poettering commented:
ok, thanks for the review! merging!
